deps/chakrashim/lib/chakra_debug.js

```diff
@@ -0,0 +1,1304 @@
+        logFunc(logTypes.API, arguments);
+      },
+      LogOutput: function(msg, obj) {
+        logFunc(logTypes.Output, arguments);
+      },
+      LogInput: function(msg, obj) {
+        logFunc(logTypes.Input, arguments);
+      },
+      LogError: function(msg, obj) {
+        logFunc(logTypes.Error, arguments);
+      }
+    };
+  })();
+
+  function CallHostFunctionNoLog(fn) {
+    var args = [].slice.call(arguments, 1);
+    var start = new Date();
+    var result = fn.apply(undefined, args);
+    var end = new Date();
+    Logger.LogAPI(`${fn.name}(${args}):${end - start}`);
+    return result;
+  }
+
+  function CallHostFunction(fn) {
+    var args = [].slice.call(arguments, 1);
+    var start = new Date();
+    var result = fn.apply(undefined, args);
+    var end = new Date();
+    Logger.LogAPI(`${fn.name}(${args}):${end - start}`, result);
+    return result;
+  }
+
+  if (!globalObject.Debug) {
+    Object.defineProperty(globalObject, 'Debug', {
+      value: {}, enumerable: false, configurable: false, writable: false
+    });
+  }
+
+  var Debug = globalObject.Debug;
+
+  Debug.ScriptBreakPointType = {
+    ScriptId: 0,
+    ScriptName: 1,
+    ScriptRegExp: 2
+  };
+
+  Debug.setListener = function(fn) { };
+
+  Debug.setBreakPoint = function(func, line, column) {
+    Logger.LogAPI(`Debug.setBreakPoint(${func},${line},${column})`);
+    var funcInfo = CallHostFunction(chakraDebug.JsDiagGetFunctionPosition,
+                                     func);
+    var bpId = -1;
+
+    if (funcInfo.scriptId >= 0) {
+      var scriptObj = DebugManager.ScriptsManager.GetScript(funcInfo.scriptId);
+      var bpLine = funcInfo.firstStatementLine + line;
+      var bpColumn = funcInfo.firstStatementColumn + column;
+      var v8Breakpoint = new V8Breakpoint(Debug.ScriptBreakPointType.ScriptId,
+                              funcInfo.scriptId, scriptObj, bpLine, bpColumn);
+      if (v8Breakpoint.Set()) {
+        bpId = v8Breakpoint.Id();
+        DebugManager.BreakpointManager.AddBreakpoint(v8Breakpoint);
+      }
+    }
+
+    return bpId;
+  };
+
+  function AddChildrens(obj) {
+    if ('display' in obj) {
+      if (!('value' in obj) || !isFinite(obj['value']) ||
+        obj['value'] == null || obj['value'] == undefined) {
+        obj['value'] = obj['display'];
+      }
+      if (!('text' in obj)) {
+        obj['text'] = obj['display'];
+      }
+      if (!('className' in obj)) {
+        obj['className'] = obj['type'];
+      }
+    }
+
+    if ('className' in obj && obj['className'] == 'Object') {
+      obj.constructorFunction = {
+        'ref': globalExecutionState.GetObjectEvalHandle()
+      };
+      obj.protoObject = {
+        'ref': globalExecutionState.GetObjectProtoEvalHandle()
+      };
+      obj.prototypeObject = {
+        'ref': globalExecutionState.GetObjectPrototypeEvalHandle()
+      };
+    }
+
+    var PROPERTY_ATTRIBUTE_HAVE_CHILDRENS = 0x1;
+    //var PROPERTY_ATTRIBUTE_READ_ONLY_VALUE = 0x2;
```

Comment:
Currently we don't use the read only attribute, but I will keep it here for clarity.

---
In reply to: [81491126](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81491126) [](ancestors = 81491126)